frontend/livepatch.lua

```diff
@@ -0,0 +1,139 @@
+                    execute("sh", fullpath, home_dir .. "/koreader", package_dir)
+                elseif mode == "file" and fullpath:match("%.lua$")
+                    and not fullpath:match("master%-patch%.lua$") then -- execute patch-files
+                    local ok, err = pcall(dofile, fullpath)
+                    if not ok then
+                        logger.warn("Live update", err)
+                        if not bare then -- Only show InfoMessage, when late during startup.
+                            -- Only developers (advanced users) will use this mechanism.
+                            -- A warning on a patch failure after an OTA update will simplify troubleshooting.
+                            local UIManager = require("ui/uimanager")
+                            local _ = require("gettext")
+                            local baseUtil = require("ffi/util")
+                            local T = baseUtil.template
+                            local InfoMessage = require("ui/widget/infomessage")
+                            UIManager:show(InfoMessage:new{text = T(_("Error loading patch:\n%1"), fullpath)})
+                        end
+                    else
+                        logger.info("Live update applied:", fullpath)
+                    end
+                elseif mode == "directory" then -- recurse deeper
+                    -- recurse into next directory
+                    runLiveUpdateTasks(fullpath, bare, migration, parent and parent .. "/" .. entry or entry)
+                end
+            end
+        end
+    end
+
+    --- This function works on the contents of `home_dir/koreader/patches` (except `master-patch.lua`)
+    function livepatch.applyPatches()
+        -- patches, scripts and migration get applied at every start of koreader, no migration here
+        local patch_dir = home_dir .. "/koreader/patches"
+        if lfs.attributes(patch_dir, "mode") == "directory" then
+            runLiveUpdateTasks(patch_dir, false)
+        end
+    end
+
+    --- This function migrates user files, executes shell scripts and patch files (except `master-patch.lua`) in
+    -- `home_dir/koreader/[scripts.afterupdate|scripts.always`.
+    -- If an older `patch.lua` is found it is moved to `patches/master-patch.lua`. This has to be done here,
+    -- as onetime_migration gets called to late.
+    function livepatch.executeScriptsAndMigrate()
+        local run_once_scripts = home_dir .. "/koreader/scripts.afterupdate"
+        if lfs.attributes(run_once_scripts, "mode") == "directory" then
+            local afterupdate_marker = package_dir .. "/koreader/afterupdate.marker"
+            if lfs.attributes(afterupdate_marker, "mode") == "file" then
+                local migrate = lfs.attributes(run_once_scripts .. "/migrate", "mode") ~= nil
+                logger.info("after-update: running", migrate and "migration" or "shell scripts")
```

Comment:
What for do you use this "/migrate/" stuff ? For your german hyphenation test file ? Anything else ?
Dunno if other users than you do use this feature - but may be we could simplify the amount of available tricks :) and just use early and last lua scripts. You could do any move/migration from a Lua script (even if shell scripts might be easier to write for that kind of thing), it would feel cleaner to me.
You could add helpers methods in this livepatch.lua so shell stuff like cp can be easily done in a lua file.

(I'm having a hard time understanding what's possible and what happens when reading this new linvepatch.lua.)

Reply:
One question concerning the execute function. Now it uses `os.execute`. Would it be of any benefit to use `os.rename` and `os.remove` where possible?